subsys/bluetooth/controller/ticker/ticker.c

```diff
@@ -536,15 +536,13 @@ void ticker_worker(void *param)
 	ticker_id_head = instance->ticker_id_head;
 
 	/* Expire all tickers within ticks_elapsed and collect ticks_expired */
-	node = &instance->nodes[0];
 	while (ticker_id_head != TICKER_NULL) {
-		struct ticker_node *ticker;
 		u32_t ticks_to_expire;
 
-		ticker = &node[ticker_id_head];
+		node = &instance->nodes[ticker_id_head];
```

Comment:
Avoid double dereferencing inside the loop, it is expensive on CPU time. Hence revert back to using an auto variable assigned outside the loop.

Reply:
In this case the first dereferencing doesn't cost anything because the offset of the "nodes" fields relative to "instance" is known at compile time and it doesn't have to be computed at run time. Therefore the two variants of code result in similar assembly.

Here is a dump of the code before my patch:
```asm
		ticker = &node[ticker_id_head];
    a666:	2328	movs	r3, #40	; 0x28
    a668:	435f	muls	r7, r3
    a66a:	9b01	ldr	r3, [sp, #4]
		if (ticks_elapsed < ticks_to_expire) {
    a66c:	9a00	ldr	r2, [sp, #0]
		ticker = &node[ticker_id_head];
    a66e:	19d9	adds	r1, r3, r7
```

And this is after the patch:

```asm
		node = &instance->nodes[ticker_id_head];
    a662:	2328	movs	r3, #40	; 0x28
    a664:	435f	muls	r7, r3
    a666:	6823	ldr	r3, [r4, #0]
		if (ticks_elapsed < ticks_to_expire) {
    a668:	9a01	ldr	r2, [sp, #4]
		node = &instance->nodes[ticker_id_head];
    a66a:	19d9	adds	r1, r3, r7
```

The instruction ```ldr r3, [r4, #0]``` loads the address of instance->nodes to r3. Since the "nodes" field is the first one the offset is #0. But if it were any other value it would be filled there since it is already known at compile time. Therefore there is no cicle penalty, only clarity in code.

Reply:
This is probably compiler/architecture dependent. I agree with Vinayak to keep the auto variable.

Reply:
The example I've given is for ARM, the only platform that BLE now compiles to. But the principle is the same for every architecture.

Reply:
It sometime back when I check the generated opcodes and the use of auto variable is an outcome of developing for multiple architectures and with no optimization in compiler options. Did you compare with CONFIG_NO_OPTIMIZATIONS=y ?

Reply:
CM0, no optimization:
```
34452         /* Expire all tickers within ticks_elapsed and collect ticks_expired */ 
34453         //node = &instance->nodes[0];                                           
34454         while (ticker_id_head != TICKER_NULL) {                                 
34455     bf04:       e048            b.n     bf98 <ticker_worker+0xe4>               
34456                 struct ticker_node *ticker;                                     
34457                 u32_t ticks_to_expire;                                          
34458                                                                                 
34459                 ticker = &instance->nodes[ticker_id_head];                      
34460     bf06:       69bb            ldr     r3, [r7, #24]                           
34461     bf08:       6819            ldr     r1, [r3, #0]                            
34462     bf0a:       231f            movs    r3, #31                                 
34463     bf0c:       18fb            adds    r3, r7, r3                              
34464     bf0e:       781a            ldrb    r2, [r3, #0]                            
34465     bf10:       0013            movs    r3, r2                                  
34466     bf12:       009b            lsls    r3, r3, #2                              
34467     bf14:       189b            adds    r3, r3, r2                              
34468     bf16:       00db            lsls    r3, r3, #3                              
34469     bf18:       18cb            adds    r3, r1, r3                              
34470     bf1a:       617b            str     r3, [r7, #20]                           
34471                                                                                 
34472                 /* Stop if ticker did not expire */                             
34473                 ticks_to_expire = ticker->ticks_to_expire;                      
```

CM0 with size optimization:
```
47106         /* Expire all tickers within ticks_elapsed and collect ticks_expired */ 
47107         //node = &instance->nodes[0];                                           
47108         while (ticker_id_head != TICKER_NULL) {                                 
47109     abe6:       2fff            cmp     r7, #255        ; 0xff                  
47110     abe8:       d007            beq.n   abfa <ticker_worker+0x3c>               
47111                 struct ticker_node *ticker;                                     
47112                 u32_t ticks_to_expire;                                          
47113                                                                                 
47114                 ticker = &instance->nodes[ticker_id_head];                      
47115     abea:       2328            movs    r3, #40 ; 0x28                          
47116     abec:       435f            muls    r7, r3                                  
47117     abee:       6823            ldr     r3, [r4, #0]                            
47118                                                                                 
47119                 /* Stop if ticker did not expire */                             
47120                 ticks_to_expire = ticker->ticks_to_expire;                      
```

CM0 size optimized upstream code:
```
47107         /* Expire all tickers within ticks_elapsed and collect ticks_expired */ 
47108         node = &instance->nodes[0];                                             
47109     abe6:       6823            ldr     r3, [r4, #0]                            
47110     abe8:       9301            str     r3, [sp, #4]                            
47111         while (ticker_id_head != TICKER_NULL) {                                 
47112     abea:       2fff            cmp     r7, #255        ; 0xff                  
47113     abec:       d007            beq.n   abfe <ticker_worker+0x40>               
47114                 struct ticker_node *ticker;                                     
47115                 u32_t ticks_to_expire;                                          
47116                                                                                 
47117                 ticker = &node[ticker_id_head];                                 
47118     abee:       2328            movs    r3, #40 ; 0x28                          
47119     abf0:       435f            muls    r7, r3                                  
47120     abf2:       9b01            ldr     r3, [sp, #4]                            
47121                                                                                 
47122                 /* Stop if ticker did not expire */                             
47123                 ticks_to_expire = ticker->ticks_to_expire;                      
```

CM0 no optimization, upstream code:
```
34452         /* Expire all tickers within ticks_elapsed and collect ticks_expired */ 
34453         node = &instance->nodes[0];                                             
34454     bf04:       69bb            ldr     r3, [r7, #24]                           
34455     bf06:       681b            ldr     r3, [r3, #0]                            
34456     bf08:       617b            str     r3, [r7, #20]                           
34457         while (ticker_id_head != TICKER_NULL) {                                 
34458     bf0a:       e048            b.n     bf9e <ticker_worker+0xea>               
34459                 struct ticker_node *ticker;                                     
34460                 u32_t ticks_to_expire;                                          
34461                                                                                 
34462                 ticker = &node[ticker_id_head];                                 
34463     bf0c:       231f            movs    r3, #31                                 
34464     bf0e:       18fb            adds    r3, r7, r3                              
34465     bf10:       781a            ldrb    r2, [r3, #0]                            
34466     bf12:       0013            movs    r3, r2                                  
34467     bf14:       009b            lsls    r3, r3, #2                              
34468     bf16:       189b            adds    r3, r3, r2                              
34469     bf18:       00db            lsls    r3, r3, #3                              
34470     bf1a:       001a            movs    r2, r3                                  
34471     bf1c:       697b            ldr     r3, [r7, #20]                           
34472     bf1e:       189b            adds    r3, r3, r2                              
34473     bf20:       613b            str     r3, [r7, #16]                           
34474                                                                                 
34475                 /* Stop if ticker did not expire */                             
34476                 ticks_to_expire = ticker->ticks_to_expire;                      
```

Reply:
Even for size-optimized compilation the output is the same. And who compiles for C0 and worries about 1 extra instruction added to 10 instructions for probably 3-4 loops? (at most).

Reply:
It is not about number of instructions, it about whether the load is done inside the while loop for every iteration or once before the loop. I may have not done a thorough analysis of the snippet I have pasted, for which I need to allocate time out of my current tasks. 

You may create a new PR with de-referencing changes, can we stick to name change in this PR?

Reply:
As the code is you already do a load inside the loop (since you access node[ticker_id_head]) but the questions is if you do an extra one (when you access ticker->node[]).

I am in no hurry for the moment so I'll wait for your review.

Reply:
> The example I've given is for ARM, the only platform that BLE now compiles to. But the principle is the same for every architecture.

Just for the record, ARM is _not_ the only architecture Zephyr BLE compiles to. I hope that's not a common assumption...

Reply:
The same principles apply for every platform. I've only tested for arm. But from what I know the upstream only builds for Nordic (which is based on arm if I am not mistaking). Are there any other platforms publicly available?